### PR TITLE
Release 2.2 - amplification model and rename xyz to dmm

### PR DESCRIPTION
**Pull request by ducquangkstn**
- Add amplification model, details in CHANGELOG.MD
- rename `xyz` to `dmm`
- add feeBps config to factory contract

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,7 +1,14 @@
...
 - The architecture with factory, pair and router is the same with uniswap model.
 
 Also change:
-- Remove price oracle feature.
\ No newline at end of file
+- Remove price oracle feature.
+### AMP model: allow user to amplify the reserve balance of a pair
+- [the white paper](https://github.com/xyzswap/xyz-protocol/blob/main/apr_v2.pdf)
+- For each pair, factory contract can create many pools with different amplification factor. 
+But there is only pair with amplification factor = 1
```
Comment (manhlx3006) [inline CHANGELOG.md]: can we have multiple pools for the same amplification factor?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,7 +1,14 @@
...
 - Fee fomula is implemented at [here](contracts/libraries/FeeFomula.sol)
 - The architecture with factory, pair and router is the same with uniswap model.
 
 Also change:
-- Remove price oracle feature.
\ No newline at end of file
+- Remove price oracle feature.
+### AMP model: allow user to amplify the reserve balance of a pair
+- [the white paper](https://github.com/xyzswap/xyz-protocol/blob/main/apr_v2.pdf)
+- For each pair, factory contract can create many pools with different amplification factor.
```
Comment (manhlx3006) [inline CHANGELOG.md]: `multiple pools with different amplification factors`

File: contracts/XYZSwapPair.sol [hunk 3]
```diff
@@ -8,26 +8,42 @@ import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
...
+        uint256 reserve0;
+        uint256 reserve1;
+        uint256 vReserve0;
+        uint256 vReserve1; // only used when isAmpPool = true
+    }
+
     uint256 public constant MINIMUM_LIQUIDITY = 10**3;
 
-    address public factory;
+    IXYZSwapFactory public factory;
```
Comment (manhlx3006) [inline contracts/XYZSwapPair.sol]: can add a comment with the reason that we don't use `immutable` for `factory` here

File: contracts/XYZSwapPair.sol [hunk 4]
```diff
@@ -8,26 +8,42 @@ import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
...
     IERC20 public token0;
     IERC20 public token1;
 
-    /// @dev uses single storage slot, accessible via getReserves
+    /// @dev uses single storage slot, accessible via getReservesData
     uint112 internal reserve0;
     uint112 internal reserve1;
     uint32 internal blockTimestampLast;
-    /// @dev reserve0 * reserve1, as of immediately after the most recent liquidity event
+    uint32 internal ampBps;
```
Comment (manhlx3006) [inline contracts/XYZSwapPair.sol]: where can we get the `ampBps` data for the pair?

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,7 +1,14 @@
-### XYZ model: change fee calculation based on reserve and volume.
+### XYZ model: change fee calculation based on reserve and volume
+- [the white paper](https://github.com/xyzswap/xyz-protocol/blob/main/xyz.pdf)
```
Comment (Anyhowclick) [inline CHANGELOG.md]: the -> The

File: contracts/XYZSwapFactory.sol [hunk 6]
```diff
@@ -1,42 +1,60 @@
...
         require(tokenA != tokenB, "XYZSwap: IDENTICAL_ADDRESSES");
         (IERC20 token0, IERC20 token1) = tokenA < tokenB ? (tokenA, tokenB) : (tokenB, tokenA);
         require(address(token0) != address(0), "XYZSwap: ZERO_ADDRESS");
-        require(getPair[token0][token1] == address(0), "XYZSwap: PAIR_EXISTS"); // single check is sufficient
-        bytes memory bytecode = type(XYZSwapPair).creationCode;
-        bytes32 salt = keccak256(abi.encodePacked(token0, token1));
-        assembly {
-            pair := create2(0, add(bytecode, 32), mload(bytecode), salt)
+        require(ampBps >= BPS, "XYZSwap: INVALID_BPS");
+        // only exist 1 non-amplificaiton pool of a pair.
```
Comment (Anyhowclick) [inline contracts/XYZSwapFactory.sol]: amplification

File: package.json [hunk 7]
```diff
@@ -13,6 +13,7 @@
     "test": "test"
   },
   "devDependencies": {
+    "@kyber.network/utils-sc": "^2.1.3",
```
Comment (Anyhowclick) [inline package.json]: Remove the carats `^`

**Review by Anyhowclick — CHANGES_REQUESTED** [review]

File: contracts/XYZSwapPair.sol [hunk 8]
```diff
@@ -90,71 +123,92 @@ contract XYZSwapPair is IXYZSwapPair, ERC20Permit, ReentrancyGuard, VolumeTrendR
...
+        uint256 amount0In = newData.reserve0 > data.reserve0 - amount0Out
+            ? newData.reserve0 - (data.reserve0 - amount0Out)
             : 0;
-        uint256 amount1In = balance1 > _reserve1 - amount1Out
-            ? balance1 - (_reserve1 - amount1Out)
+        uint256 amount1In = newData.reserve1 > data.reserve1 - amount1Out
+            ? newData.reserve1 - (data.reserve1 - amount1Out)
             : 0;
         require(amount0In > 0 || amount1In > 0, "XYZSwap: INSUFFICIENT_INPUT_AMOUNT");
+        uint256 feeInPrecision = verifyBalanceAndUpdateEma(
```
Comment (manhlx3006) [inline contracts/XYZSwapPair.sol]: should we do this instead?
`uint256 feeInPrecision = isAmpPool ? verifyBalanceAndUpdateEma() : verifyBalanceAndUpdateEma()`

Comment (ducquangkstn) [reply contracts/XYZSwapPair.sol]: I try, actually it looks uglier

File: contracts/XYZSwapPair.sol (hunk 4 above)
Comment (ducquangkstn) [reply contracts/XYZSwapPair.sol]: change to public

File: contracts/XYZSwapPair.sol (hunk 3 above)
Comment (ducquangkstn) [reply contracts/XYZSwapPair.sol]: Ok

File: CHANGELOG.md (hunk 2 above)
Comment (ducquangkstn) [reply CHANGELOG.md]: Updated

File: CHANGELOG.md (hunk 1 above)
Comment (ducquangkstn) [reply CHANGELOG.md]: Updated

File: contracts/XYZSwapPair.sol (hunk 8 above)
Comment (manhlx3006) [reply contracts/XYZSwapPair.sol]: well I think we could save some gas by checking value of `isAmpPool1` only once, but maybe just a few gas

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,7 +1,25 @@
-### XYZ model: change fee calculation based on reserve and volume.
+### Dynamic fee model: change fee calculation based on reserve and volume
+- [The white paper](https://github.com/xyzswap/xyz-protocol/blob/main/xyz.pdf)
```
Comment (Anyhowclick) [inline CHANGELOG.md]: does the link need to be updated since we aren't calling it XYZ anymore?

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,7 +1,25 @@
...
 - Ema volume is implemented at [here](contracts/VolumeTrendRecorder.sol)
 - Fee fomula is implemented at [here](contracts/libraries/FeeFomula.sol)
 - The architecture with factory, pair and router is the same with uniswap model.
 
 Also change:
-- Remove price oracle feature.
\ No newline at end of file
+- Remove price oracle feature.
+### AMP model: allow user to amplify the reserve balance of a pair
+- [The white paper](https://github.com/xyzswap/xyz-protocol/blob/main/apr_v2.pdf)
```
Comment (Anyhowclick) [inline CHANGELOG.md]: here also...

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,7 +1,25 @@
...
+- For each pair, factory contract can create different pools with the same pair of token.   
+Because the price range depends on the initial price, we allow users can create different pools with the same amplification factor.   
+But there is only pair with amplification factor = 1
+- Router:
+    - When users swap from A->B->C, users must specify which pools they choose to swap with at router contract.   
+Router contract will also check if these pools are created from factory contract.
+    - For add/remove liquidity, users must specify which pools they choose to add/remove liquidity.   
+To add liquidity to a new pool, users would use function `addLiquidityNewPool` and `addLiquidityNewPoolETH`  
+- Change in fee:
+  - We divide pair into 4 categories, the higher amplification will inidicate that the more stable price, so we use lower base fee
```
Comment (Anyhowclick) [inline CHANGELOG.md]: - We divide token pairs into 4 categories. In general, higher amplification factors used indicate greater price stability, so we use lower base fees.

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,7 +1,25 @@
...
+    - Similar asserts pairs(amplifacationFactor > 20): baseFee = 4 bps
```
Comment (Anyhowclick) [inline CHANGELOG.md]: asset pairs (amplificationFactor > 20)

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,7 +1,25 @@
...
+    - Strongly correlated assets pairs (20 >= amplicationFactor > 5): baseFee = 10 bps
```
Comment (Anyhowclick) [inline CHANGELOG.md]: amplificationFactor

File: CHANGELOG.md [hunk 14]
```diff
@@ -1,7 +1,25 @@
...
+    - Correlated assets pairs (5 >= amplificationFactor > 2): baseFee = 20 bps
+    - Uncorrelated assets pairs: (amplifactionFactor <= 2): baseFee = 30 bps
+  - For network fee: we keep to mechanism from uniswap but let `FeeToSetter` can set fee from network from up to 20% grow of the pool (this number of uniswap is 1/6)
```
Comment (Anyhowclick) [inline CHANGELOG.md]: For network fee: we keep the same mechanism as Uniswap, but have `FeeToSetter` to set network fees up to 20% growth of the pool (Uniswap uses 1/6).

File: README.md [hunk 15]
```diff
@@ -1,10 +1,10 @@
-# XYZSwap
+# Dynamic automate market maker
```
Comment (Anyhowclick) [inline README.md]: Dynamic Automated Market Maker

File: README.md [hunk 16]
```diff
@@ -1,10 +1,10 @@
...
 ## Introduction
 [![built-with openzeppelin](https://img.shields.io/badge/built%20with-OpenZeppelin-3677FF)](https://docs.openzeppelin.com/)
 [![Build Status](https://api.travis-ci.com/xyzswap/smart-contracts.svg?branch=master&status=passed)](https://travis-ci.com/github/KyberNetwork/kyber_reserves_sc)
 
 This repository contains the xyzswap smart contracts.
-For more details, please visit [the white paper](https://github.com/xyzswap/xyz-protocol/blob/main/xyz.pdf) and our [change log](CHANGELOG.md) (compared to uniswap)
+For more details, please visit the white paper([dynamic fee](https://github.com/xyzswap/xyz-protocol/blob/main/xyz.pdf)  and [amplfication algorithm](https://github.com/xyzswap/xyz-protocol/blob/main/apr_v2.pdf))and our [change log](CHANGELOG.md) (compared to uniswap)
```
Comment (Anyhowclick) [inline README.md]: link updates (if necessary)

File: contracts/DMMPool.sol [hunk 17]
```diff
@@ -0,0 +1,375 @@
...
+    IDMMFactory public factory;
+    IERC20 public token0;
+    IERC20 public token1;
+
+    /// @dev uses single storage slot, accessible via getReservesData
+    uint112 internal reserve0;
+    uint112 internal reserve1;
+    uint32 internal blockTimestampLast;
+    uint32 public ampBps;
+    /// @dev addition param only when pool is amp
```
Comment (Anyhowclick) [inline contracts/DMMPool.sol]: /// @dev additional params used only when amplification factor > 1

File: contracts/DMMPool.sol [hunk 18]
```diff
@@ -0,0 +1,375 @@
...
+        factory = IDMMFactory(msg.sender);
+    }
+
+    // called once by the factory at time of deployment
+    function initialize(
+        IERC20 _token0,
+        IERC20 _token1,
+        uint32 _ampBps
+    ) external {
+        require(msg.sender == address(factory), "DMM: FORBIDDEN"); // sufficient check
```
Comment (Anyhowclick) [inline contracts/DMMPool.sol]: can remove the comment, doesn't make me feel any safer lol

File: contracts/DMMPool.sol [hunk 19]
```diff
@@ -0,0 +1,375 @@
...
+        token0 = _token0;
+        token1 = _token1;
+        ampBps = _ampBps;
+    }
+
+    /// @dev this low-level function should be called from a contract
+    ///                 which performs important safety checks
```
Comment (Anyhowclick) [inline contracts/DMMPool.sol:79]: too many whitespaces?

File: contracts/DMMPool.sol [hunk 20]
```diff
@@ -0,0 +1,375 @@
...
+        uint256 beforeReserve0,
+        uint256 beforeReserve1,
+        uint256 afterReserve0,
+        uint256 afterReserve1
+    ) internal virtual returns (uint256 feeInPrecision) {
+        // volume = beforeReserve0 * amount1In / beforeReserve1 + amount0In (normalized into amount in token 0)
+        uint256 volume = beforeReserve0.mul(amount1In).div(beforeReserve1).add(amount0In);
+        uint256 rFactorInPrecision = recordNewUpdatedVolume(block.number, volume);
+        feeInPrecision = getFinalFee(FeeFomula.getFee(rFactorInPrecision), ampBps);
+        //verify balance update is match with fomula
```
Comment (Anyhowclick) [inline contracts/DMMPool.sol]: // verify balance update matches with formula

File: contracts/DMMPool.sol [hunk 21]
```diff
@@ -0,0 +1,375 @@
...
+        blockTimestampLast = blockTimestamp;
+        if (isAmpPool) {
+            assert(data.vReserve0 >= data.reserve0 && data.vReserve1 >= data.reserve1); // never happen
+            vReserve0 = safeUint112(data.vReserve0);
+            vReserve1 = safeUint112(data.vReserve1);
+        }
+        emit Sync(data.vReserve0, data.vReserve1, data.reserve0, data.reserve1);
+    }
+
+    /// @dev if fee is on, mint liquidity equivalent to 1/6th of the growth in sqrt(k)
```
Comment (Anyhowclick) [inline contracts/DMMPool.sol]: /// @dev if fee is on, mint liquidity equivalent to configured fee of the growth in sqrt(k)

File: contracts/DMMPool.sol [hunk 22]
```diff
@@ -0,0 +1,375 @@
...
+        require(liquidity > 0, "DMM: INSUFFICIENT_LIQUIDITY_MINTED");
+        _mint(to, liquidity);
+
+        _update(isAmpPool, _data);
+        if (feeOn) kLast = getK(isAmpPool, _data);
+        emit Mint(msg.sender, amount0, amount1);
+    }
+
+    /// @dev this low-level function should be called from a contract
+    ///           which performs important safety checks
```
Comment (Anyhowclick) [inline contracts/DMMPool.sol]: just checking, the indent here is as expected...?

File: contracts/DMMPool.sol [hunk 23]
```diff
@@ -0,0 +1,375 @@
...
+    }
+
+    /// @dev force reserves to match balances
+    function sync() external override nonReentrant {
+        (bool isAmpPool, ReserveData memory data) = getReservesData();
+        bool feeOn = _mintFee(isAmpPool, data);
+        ReserveData memory newData;
+        newData.reserve0 = IERC20(token0).balanceOf(address(this));
+        newData.reserve1 = IERC20(token1).balanceOf(address(this));
+        // update virtual reserves if this is amp pool.
```
Comment (Anyhowclick) [inline contracts/DMMPool.sol]: can remove `.`

File: contracts/DMMFactory.sol [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+    function createPair(
+        IERC20 tokenA,
+        IERC20 tokenB,
+        uint32 ampBps
+    ) external override returns (address pair) {
+        require(tokenA != tokenB, "DMM: IDENTICAL_ADDRESSES");
+        (IERC20 token0, IERC20 token1) = tokenA < tokenB ? (tokenA, tokenB) : (tokenB, tokenA);
+        require(address(token0) != address(0), "DMM: ZERO_ADDRESS");
+        require(ampBps >= BPS, "DMM: INVALID_BPS");
+        // only exist 1 non-amplification pool of a pair.
```
Comment (Anyhowclick) [inline contracts/DMMFactory.sol]: // only 1 unamplified (or non-amplified) pool can exist

Comment (Anyhowclick) [reply contracts/DMMFactory.sol]: unamplified because that was what was used in the whitepaper

File: contracts/libraries/DMMLibrary.sol [hunk 25]
```diff
@@ -0,0 +1,167 @@
...
+        internal
+        pure
+        returns (IERC20 token0, IERC20 token1)
+    {
+        require(tokenA != tokenB, "DMMLibrary: IDENTICAL_ADDRESSES");
+        (token0, token1) = tokenA < tokenB ? (tokenA, tokenB) : (tokenB, tokenA);
+        require(address(token0) != address(0), "DMMLibrary: ZERO_ADDRESS");
+    }
+
+    /// @dev fetch the reserves and fee for a pair, used for trading purpose
```
Comment (Anyhowclick) [inline contracts/libraries/DMMLibrary.sol]: trading purposes

**Review by Anyhowclick — CHANGES_REQUESTED** [review]

File: contracts/libraries/DMMLibrary.sol (hunk 25 above)
Comment (ducquangkstn) [reply contracts/libraries/DMMLibrary.sol]: Change in #29

File: contracts/DMMFactory.sol (hunk 24 above)
Comment (ducquangkstn) [reply contracts/DMMFactory.sol]: change in #29 nonAmp -> unamplified in every places

File: contracts/DMMPool.sol (hunk 23 above)
Comment (ducquangkstn) [reply contracts/DMMPool.sol]: change in #29

File: contracts/DMMPool.sol (hunk 21 above)
Comment (ducquangkstn) [reply contracts/DMMPool.sol]: change in #29

File: CHANGELOG.md (hunk 14 above)
Comment (ducquangkstn) [reply CHANGELOG.md]: change in #29

File: README.md (hunk 15 above)
Comment (ducquangkstn) [reply README.md]: change in #29

File: README.md (hunk 16 above)
Comment (ducquangkstn) [reply README.md]: change in #29

File: CHANGELOG.md (hunk 13 above)
Comment (ducquangkstn) [reply CHANGELOG.md]: change in #29

File: CHANGELOG.md (hunk 12 above)
Comment (ducquangkstn) [reply CHANGELOG.md]: change in #29

File: CHANGELOG.md (hunk 11 above)
Comment (ducquangkstn) [reply CHANGELOG.md]: change in #29

File: contracts/DMMPool.sol (hunk 18 above)
Comment (ducquangkstn) [reply contracts/DMMPool.sol]: change in #29

File: contracts/DMMPool.sol (hunk 20 above)
Comment (ducquangkstn) [reply contracts/DMMPool.sol]: change in #29

File: contracts/DMMPool.sol (hunk 19 above)
Comment (ducquangkstn) [reply contracts/DMMPool.sol:79]: change in #29 
Aw, not sure about our standard for multiple line comments

File: contracts/DMMPool.sol (hunk 17 above)
Comment (ducquangkstn) [reply contracts/DMMPool.sol]: change in #29

File: CHANGELOG.md (hunk 9 above)
Comment (ducquangkstn) [reply CHANGELOG.md]: change in #29

File: CHANGELOG.md (hunk 10 above)
Comment (ducquangkstn) [reply CHANGELOG.md]: change in #29

File: CHANGELOG.md [hunk 26]
```diff
@@ -1,25 +1,25 @@
...
 - Change in fee:
-  - We divide pair into 4 categories, the higher amplification will inidicate that the more stable price, so we use lower base fee
-    - Similar asserts pairs(amplifacationFactor > 20): baseFee = 4 bps
-    - Strongly correlated assets pairs (20 >= amplicationFactor > 5): baseFee = 10 bps
-    - Correlated assets pairs (5 >= amplificationFactor > 2): baseFee = 20 bps
-    - Uncorrelated assets pairs: (amplifactionFactor <= 2): baseFee = 30 bps
-  - For network fee: we keep to mechanism from uniswap but let `FeeToSetter` can set fee from network from up to 20% grow of the pool (this number of uniswap is 1/6)
\ No newline at end of file
+  - We divide token pools into 4 categories. In general, the higher amplification used indicate greater price stability, so we use lower base fees.
+    - Similar asserts pools(amplificationFactor > 20): baseFee = 4 bps
```
Comment (Anyhowclick) [inline CHANGELOG.md]: assets

File: contracts/interfaces/IDMMLiquidityRouter.sol [hunk 27]
```diff
@@ -59,7 +59,7 @@ interface IDMMLiquidityRouter {
 
     function addLiquidityETH(
         IERC20 token,
-        address pair,
+        address,
```
Comment (Anyhowclick) [inline contracts/interfaces/IDMMLiquidityRouter.sol]: address pool?

**Review by Anyhowclick — CHANGES_REQUESTED** [review]

**Review by Anyhowclick — APPROVED** [review]

**Review by manhlx3006 — APPROVED** [review]

**Pull request merged**